db-service/test/cqn4sql/with-parameters.test.js

```diff
@@ -8,32 +8,191 @@ const { SELECT } = require('@sap/cds/lib/ql/cds-ql')
+          where exists (
+            SELECT 1 from Books as Books where Books.author_ID = author.ID
+          )
+      `
+      expect(cqn4sql(query, model)).to.deep.equal(expected)
+    })
+    it('where exists shortcut', () => {
+      const query = CQL`SELECT from Books { ID } where exists author(P1: 1, P2: 2)`
+      const expected = CQL`
+        SELECT from Books as Books { Books.ID }
+          where exists (
+            SELECT 1 from Authors(P1: 1, P2: 2) as author where author.ID = Books.author_ID
+          )
+      `
+      expect(cqn4sql(query, model)).to.deep.equal(expected)
+    })
+    it('where exists shortcut w/o params', () => {
+      const query = CQL`SELECT from Books { ID } where exists author`
+      const expected = CQL`
+        SELECT from Books as Books { Books.ID }
+          where exists (
+            SELECT 1 from Authors(P1: dummy) as author where author.ID = Books.author_ID
+          )
+      `
+      // manually remove the param from argument list because compiler does not allow empty args for cqn
+      expected.SELECT.where[1].SELECT.from.ref[0].args = {}
+      expect(cqn4sql(query, model)).to.deep.equal(expected)
+    })
+  })
+
+  describe('expand subqueries', () => {
+    it('expand with params', () => {
+      const query = CQL`SELECT from Books {
+        author(P1: 1, P2: 2) { ID }
+      }`
+      const expected = CQL`SELECT from Books as Books {
+        (
+          SELECT from Authors(P1: 1, P2: 2) as author {
+            author.ID
+          } where Books.author_ID = author.ID
+        ) as author
+      }`
+      expect(JSON.parse(JSON.stringify(cqn4sql(query, model)))).to.deep.equal(expected)
+    })
+    it('expand on parameterized entity without args', () => {
+      const query = CQL`SELECT from Books {
+        author { ID }
+      }`
+      const expected = CQL`SELECT from Books as Books {
+        (
+          SELECT from Authors(P1: dummy) as author {
+            author.ID
+          } where Books.author_ID = author.ID
+        ) as author
+      }`
+      // manually remove the param from argument list because compiler does not allow empty args for cqn
+      expected.SELECT.columns[0].SELECT.from.ref[0].args = {}
+      expect(JSON.parse(JSON.stringify(cqn4sql(query, model)))).to.deep.equal(expected)
+    })
   })
-  // will be done in another change
-  it.skip('select from view with param and join with normal entity', () => {
-    // currently only possible with cds-compiler beta-mode,
-    // as the view with params does not yet support associations
-    const query = cqn4sql(SELECT.from('PBooks(P1: 1, P2: 2)').columns('author.name as author'), model)
-    const expected = CQL`SELECT FROM PBooks(P1: 1, P2: 2) as PBooks
-                         left join Authors as author on author.ID = PBooks.author_ID {
-                          author.name as author
-                         }`
-    expect(query).to.deep.equal(expected)
+
+  describe('subqueries', () => {
```

Comment:
Does HANA allow a subquery at this location?

Reply:
It doesnt work on my HANA Cloud:

```sql
-- generated by cds-compiler version 4.7.7
CREATE TABLE Books (
  ID INTEGER NOT NULL,
  PRIMARY KEY(ID)
);

-- generated by cds-compiler version 4.7.7
CREATE VIEW P(IN P1 INTEGER) AS SELECT
  Books_0.ID
FROM Books AS Books_0
WHERE :P1 = Books_0.ID;

INSERT INTO Books (ID) VALUES (1);

SELECT * FROM P((SELECT 1 FROM DUMMY));
```

```
SQL Error [7] [HY000]: SAP DBTech JDBC: [7] (at 16): feature not supported: cannot support scalar subquery as input of parameterized sql view: line 1 col 17 (at pos 16)
```